src/ffi/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,127 @@
+use ffi_helpers::null_pointer_check;
+use libc::c_double;
+
+use crate::spatialmath::{quaternion::Quaternion, vector3::Vector3};
+
+/// The FFI interface for Quaternion functions and initialization. All functions are
+/// meant to be called externally from other languages
+
+/// Initialize a quaternion from raw components and retrieve the C pointer
+/// to its address 
+#[no_mangle]
+pub extern "C" fn new_quaternion(real: f64, i: f64, j: f64, k: f64) -> *mut Quaternion {
+    Quaternion::new(real, i, j, k).to_raw_pointer()
+}
+
+/// Initialize a quaternion from a real part and a C pointer to a Vector3
+/// and retrieve the C pointer to its address 
+#[no_mangle]
+pub unsafe extern "C" fn new_quaternion_from_vector(
+    real: f64, imag_ptr: *const Vector3
+) -> *mut Quaternion {
+    null_pointer_check!(imag_ptr);
+    let imag = *imag_ptr;
+    Quaternion::new_from_vector(real, imag).to_raw_pointer()
+}
+
+/// Free memory at the address of the quaternion pointer. Outer processes
+/// that work with Quaternions via the FFI interface MUST remember 
+/// to call this function when finished with a quaternion
+#[no_mangle]
+pub unsafe extern "C" fn free_quaternion_memory(ptr: *mut Quaternion) {
+    if ptr.is_null() {
+        return;
+    }
+    Box::from_raw(ptr);
+}
+
+/// Get the components of a quaternion as a list of C doubles, the order of the
+/// components will be (real, i, j, k)
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_get_components(quat_ptr: *const Quaternion) -> *const c_double {
+    null_pointer_check!(quat_ptr);
+    let quat = *quat_ptr;
+    let components: [c_double;4] = [quat.real, quat.i, quat.j, quat.k];
```

Comment:
[q] Is this output in a double to be able to talk to rdk/ another library? Ie, is this the over the wire message at the end of the day?

Reply:
yes